packages/flutter/lib/src/material/app_bar.dart

```diff
@@ -709,13 +712,62 @@ class _FloatingAppBarState extends State<_FloatingAppBar> {
 
     // When a scroll stops, then maybe snap the appbar into view.
     // Similarly, when a scroll starts, then maybe stop the snap animation.
+
+    // This animation is typically driven  by the RenderSliverFloatingPersistentHeader,
+    // which mocks a change in the scroll position (the _effectiveScrollOffset)
+    // to float the header in and out.
+    // When used in a NestedScrollView however, the header is not within the
+    // same scroll view, so the scroll position must be changed to achieve the
+    // same effect. In this case, indicated by
+    // FloatingHeaderSnapConfiguration.nestedSnap, we drive the animation on the
+    // scroll position here instead.
+    if (widget.snapConfiguration != null && widget.snapConfiguration.nestedSnap) {
+      _nestedAnimation();
+    } else {
+      _defaultAnimation();
+    }
+  }
+
+  void _defaultAnimation() {
     final RenderSliverFloatingPersistentHeader header = _headerRenderer();
     if (_position.isScrollingNotifier.value)
       header?.maybeStopSnapAnimation(_position.userScrollDirection);
     else
       header?.maybeStartSnapAnimation(_position.userScrollDirection);
```

Comment:
As discussed over VC: Can you remind my why just calling `maybeStartSnapAnimation` doesn't work when `floatHeaderSlivers` is set to true?

Reply:
This is due to the fact that as the scroll offset is changing, the snap animation can be broken when animating closed. When the app bar goes to animate closed, it is animating from the `_effectiveScrollOffset` to the `maxExtent` of the `RenderSliverFloatingPersistentHeader`, however if the actual `scrollOffset` is less than the `maxExtent` the animation gets stuck. In the case of `NestedScrollView`, the `scrollOffset` will always <= `maxExtent` since the `maxExtent` is actually the `maxScrollExtent` of the entire scrollable. Usually in a `CustomScrollView`, the `scrollOffset` is much greater since the slivers that follow the app bar are part of the same scrollable.

For example, if you checkout this branch and run the sample without the `nestedSnap` flag on the `SliverAppBar` you can reproduce this:
1. I scroll away an outer nested floating app bar fully out of view, and at its maxExtent
  `scrollOffset: 224.0, _effectiveScrollOffset: 224.0`
2. I scroll in reverse, floating the app bar __into view__ by ~90.0 pixels, 
  `scrollOffset:  132.0, _effectiveScrollOffset: 132.0`
3. Once I release the drag, the animation is triggered, animating from the `_effectiveScrollOffset --> 0.0`, revealing the entire app bar. The actual scrollOffset does not change during the animation, so when the animation completes we have 
  `scrollOffset: 132.0, _effectiveScrollOffset: 0.0`. 
4. Now, I scroll forward, starting to float some of the app bar __out of view__ by ~10.0 pixels
  `scrollOffset: 142.0, _effectiveScrollOffset: 10.0`
5. Once I release the drag, animation will be triggered to close the app bar, sliding it out of view, animating from the `_effectiveScrollOffset --> maxExtent (224.0)`. The scrollOffset does not change during this animation, so the animation gets stuck when it reaches the actual scrollOffset of 142.0, never reaching 224.0, leaving part of the app bar still exposed. After animating:
  `scrollOffset: 142.0, _effectiveScrollOffset: 142.0`

Reply:
Discussed this further over VC today...

We talked about landing in the else block here (on master branch):
https://github.com/flutter/flutter/blob/74a2e6a5e06332ae8689b077b5c37f4213fde997/packages/flutter/lib/src/rendering/sliver_persistent_header.dart#L574
but it's actually breaking down prior to that.

The animation gets stuck because of this line (on master branch): https://github.com/flutter/flutter/blob/74a2e6a5e06332ae8689b077b5c37f4213fde997/packages/flutter/lib/src/rendering/sliver_persistent_header.dart#L572

The _effectiveScrollOffset gets clamped to be within the bounds of the current _real_ scroll offset during layout, so when trying to animate to (example) 224.0, it can't exceed an actual scroll offset of (example) 132.0.

Reply:
Update: We discussed using `animateTo` instead of `setPixels`. I used `setPixels` instead of `animateTo` since this specific animation is triggered by the listener on the scrolling activity. Calling `animateTo` for this animation results in an infinite loop as it just re-triggers itself.